cli/main.go

```diff
@@ -191,50 +194,69 @@ func main() {
 	if opts.Config == "" {
 		fatal(errors.New("config path cannot be empty"))
 	}
-
 	configDir := opts.Config
-	ctx = config.WithDir(ctx, configDir)
+	config.WithDir(configDir)
 
 	currentContext := determineCurrentContext(opts.Context, configDir, opts.Hosts)
+	apicontext.WithCurrentContext(currentContext)
 
 	s, err := store.New(configDir)
 	if err != nil {
 		mobycli.Exec(root)
 	}
+	store.WithContextStore(s)
 
 	ctype := store.DefaultContextType
 	cc, _ := s.Get(currentContext)
 	if cc != nil {
 		ctype = cc.Type()
 	}
 
+	service, err := getBackend(ctype, configDir, opts)
+	if err != nil {
+		fatal(err)
+	}
+	backend.WithBackend(service)
+
 	root.AddCommand(
 		run.Command(ctype),
 		compose.Command(ctype),
 		volume.Command(ctype),
 	)
-	if ctype == store.DefaultContextType || ctype == store.LocalContextType {
-		cnxOptions := cliflags.CommonOptions{
-			Context:   opts.Context,
-			Debug:     opts.Debug,
-			Hosts:     opts.Hosts,
-			LogLevel:  opts.LogLevel,
-			TLS:       opts.TLS,
-			TLSVerify: opts.TLSVerify,
+
+	if err = root.ExecuteContext(ctx); err != nil {
+		handleError(ctx, err, ctype, currentContext, cc, root)
+	}
+	metrics.Track(ctype, os.Args[1:], metrics.SuccessStatus)
+}
+
+func getBackend(ctype string, configDir string, opts cliopts.GlobalOpts) (backend.Service, error) {
+	switch ctype {
```

Comment:
why have this specific case for local / default backend, and not also use the init() mechanism, same as for ACI/ECS (using singletons to initialise cliOptions, same as for configDir, etc)
This would also move the api instanciation in the local backend code, that seems more logical

Reply:
because local backend require a docker.API which we build by CLI options (otherwise need to create another singleton for this purpose).

Reply:
OK, as discussed it will be easier to merge this and we can try follow-up improvements on top of this work